From ffe38430e414f5f1f6f80c8f98c85760233c1120 Mon Sep 17 00:00:00 2001
From: vladimir.a.ivanov <vladimir.a.ivanov@intel.com>
Date: Fri, 25 Oct 2013 16:40:53 +0700
Subject: Dalvik: force stack alignment for external calls

BZ: 146860

Force stack alignment on 16 bytes for external calls.

Category: device-enablement
Domain: AOSP-Dalvik-Runtime
Origin: internal
Upstream-Candidate: yes

Change-Id: I6ea1890a536b7125bdabf07b964d00e70ba88bf6
Orig-MCG-Change-Id: I2827a1378f370a177fc7c0217a9bf9874b4b30fe
Signed-off-by: vladimir.a.ivanov <vladimir.a.ivanov@intel.com>
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
---
 vm/arch/x86/Call386ABI.S |    5 +++++
 1 files changed, 5 insertions(+), 0 deletions(-)

diff --git a/vm/arch/x86/Call386ABI.S b/vm/arch/x86/Call386ABI.S
index 766aff7..7140c28 100644
--- a/vm/arch/x86/Call386ABI.S
+++ b/vm/arch/x86/Call386ABI.S
@@ -94,6 +94,11 @@ dvmPlatformInvoke:
     pushl    %esi
     pushl    %ebx
     subl     $12,%esp
+
+/* force stack alignment. In the case of already aligned SP do nothing */
+    subl     $15,%esp
+    andl     $0xfffffff0,%esp
+
 /* For 386 ABI, argInfo hints should always be valid.  Abort if not. */
     movl     16(%ebp),%ebx
     testl    %ebx,%ebx
-- 
1.7.4.1

